Enable multiarch builds for pr-creator image

chandramerla opened the pull request:
**What this PR does / why we need it**:
This PR updates base image (golang) version and Go version to newer versions and mainly enables multi-arch images for pr-creator.

Multi-arch images for pr-creator are needed as now with #3566 we want to move PR creation part into periodic-kubevirtci-bump-centos-base-s390x. So that after x86 centos images are built, followed by s390x builds and manifest-list, we will create the PR in the s390x machine. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
```release-note
NONE
```

chandramerla commented:
/test ?

brianmcarey reviewed (commented):
Can you update the presubmit as well? 
https://github.com/kubevirt/project-infra/blob/0f240214c174c09d8f83ded838c2d17732caa185/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-presubmits.yaml#L422

Then I will be able to run rehearsals to test it.

chandramerla commented:
> Can you update the presubmit as well?
> 
> https://github.com/kubevirt/project-infra/blob/0f240214c174c09d8f83ded838c2d17732caa185/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-presubmits.yaml#L422
> 
> Then I will be able to run rehearsals to test it.

@brianmcarey I've updated project-infra-presubmits.yaml

brianmcarey commented:
/rehearse

chandramerla commented:
/retest

brianmcarey commented:
/rehearse

brianmcarey commented:
/rehearse

chandramerla commented:
Last run of job rehearsal-build-pr-creator-image failed as arm64 build output path for labels-checker ie., ./bazel-out/aarch64-fastbuild/bin/robots/cmd/labels-checker/labels-checker_/labels-checker
is different from amd64 path ie., ./bazel-out/k8-fastbuild/bin/robots/cmd/labels-checker/labels-checker_/labels-checker

So, Now I've updated bazel based build of labels-checker with go native build and tested it locally and multi-arch build of pr-creator is successful. As bazel is not supported in s390x, I've replaced bazel based build with go native build, there the above problem won't be there anyway as it generates the output in the given path across archs.

Could someone please trigger reharse job again?

brianmcarey commented:
/rehearse

brianmcarey commented:
/test all

brianmcarey commented on images/pr-creator/Dockerfile:
```diff
@@ -15,10 +15,13 @@ RUN set -x && \
-    bazel build //robots/cmd/labels-checker:labels-checker && \
-    cp ./bazel-out/k8-fastbuild/bin/robots/cmd/labels-checker/labels-checker_/labels-checker /usr/local/bin && \
+    cd robots/cmd/labels-checker && \
+    /usr/local/bin/runner.sh /bin/sh -ce 'go build -o labels-checker' && \
```
Can you build with something like 
```suggestion
    /usr/local/bin/runner.sh /bin/sh -ce 'go build ./robots/cmd/labels-checker' && \
```
that would allow us to remove the two `cd` lines.

chandramerla replied:
@brianmcarey Thanks for the feedback.
I've addressed this replacing all code incl. and inbetween the two cd lines as below and tested.
`/usr/local/bin/runner.sh /bin/sh -ce 'go build -o /usr/local/bin/labels-checker ./robots/cmd/labels-checker' && \`

brianmcarey commented:
/test all

brianmcarey commented on github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-postsubmits.yaml:
```diff
@@ -467,16 +467,9 @@ postsubmits:
               privileged: true
-            resources:
-              requests:
-                memory: "16Gi"
```
Can you please just re-add the memory requests here ? This just ensures that is gets enough RAM on the node.

chandramerla replied:
As this is now a multi-arch cross build on x86 similar to [publish-bootstrap-image](https://github.com/kubevirt/project-infra/blob/5ec81c03232510824d4c163e47038304a3163178/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-postsubmits.yaml#L229-L233), shall I add same resources as defined there ? ie.,

```
             resources:
              requests:
                memory: "29Gi"
              limits:
                memory: "29Gi"
```

brianmcarey replied:
Yes that should be good

brianmcarey commented on github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-presubmits.yaml:
```diff
@@ -419,16 +419,9 @@ presubmits:
             privileged: true
-          resources:
-            requests:
-              memory: "16Gi"
```
Can you please just re-add the memory requests here ? This just ensures that is gets enough RAM on the node.

brianmcarey replied:
> @brianmcarey You mean only memory requests ie., ` resources: requests: memory: "16Gi"`
> 
> without adding back the cpu requests and memory+cpu limits ?
> 
Yes without the cpu requests and memory+cpu limits - thanks

brianmcarey replied:
> 
> P.S: As we discussed earlier in DM, I will add -a flag to ./publish_multiarch_image.sh once testing is complete as this is pre-submit job.

I am ok with leaving the presubmit as is for now - this image doesn't get updated that often so doing the full build of the three images in the presubmit should not be too costly.

chandramerla replied:
Better I will add -a. That way I can keep the same resource requests 16Gi, than changing to 29Gi, as multi-arch build for all archs needs more resources. WDYT?

chandramerla replied:
For now I'm leaving this as is (without -a) and adding 29Gi for memory.

Later before merge I can change - to add -a and keeping only memory request to 16Gi.

brianmcarey commented:
/rehearse

brianmcarey reviewed (approved):
/approve
